### ISSUE TITLE
How to make btcpayserver-docker installation listen on ports 8333 (bitcoin) & 9735 (lightning) (on Google Cloud GCP CE)

### ISSUE THREAD

**Issue by maltokyo** [btcpayserver/btcpayserver-docker#161]
I would like to connect directly to ports 8333 (full node, receive connections from the internet) and open 9735 to the internet as well.

When I check what ports are listening with `ss -plant | grep LISTEN` - I get this:

> LISTEN      0        128         127.0.0.53%lo:53                0.0.0.0:*       users:(("systemd-resolve",pid=525,fd=13))                                      
> LISTEN      0        128               0.0.0.0:22                0.0.0.0:*       users:(("sshd",pid=977,fd=3))                                                  
> LISTEN      0        128                     *:9735                    *:*       users:(("docker-proxy",pid=24984,fd=4))                                        
> LISTEN      0        128                     *:80                      *:*       users:(("docker-proxy",pid=25495,fd=4))                                        
> LISTEN      0        128                  [::]:22                   [::]:*       users:(("sshd",pid=977,fd=4))                                                  
> LISTEN      0        128                     *:443                     *:*       users:(("docker-proxy",pid=25483,fd=4)) 

Two issues, 
1. bitcoind is not listening at all on port 8333
2. lightning seems to be listening on port 9735, and even though port forwarding from the router to the server is turned on for both ports, it doesn't seem to receive connections from the outside world.

`iptables -L` shows this, but all of these DOCKER chains, I have no idea how to work with them:

> Chain INPUT (policy ACCEPT)
> target     prot opt source               destination         
> sshguard   all  --  anywhere             anywhere            
> 
> Chain FORWARD (policy DROP)
> target     prot opt source               destination         
> DOCKER-USER  all  --  anywhere             anywhere            
> DOCKER-ISOLATION-STAGE-1  all  --  anywhere             anywhere            
> ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
> DOCKER     all  --  anywhere             anywhere            
> ACCEPT     all  --  anywhere             anywhere            
> ACCEPT     all  --  anywhere             anywhere            
> ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
> DOCKER     all  --  anywhere             anywhere            
> ACCEPT     all  --  anywhere             anywhere            
> ACCEPT     all  --  anywhere             anywhere            
> 
> Chain OUTPUT (policy ACCEPT)
> target     prot opt source               destination         
> 
> Chain DOCKER (2 references)
> target     prot opt source               destination         
> ACCEPT     tcp  --  anywhere             172.28.0.8           tcp dpt:9735
> ACCEPT     tcp  --  anywhere             172.28.0.4           tcp dpt:https
> ACCEPT     tcp  --  anywhere             172.28.0.4           tcp dpt:http
> 
> Chain DOCKER-ISOLATION-STAGE-1 (1 references)
> target     prot opt source               destination         
> DOCKER-ISOLATION-STAGE-2  all  --  anywhere             anywhere            
> DOCKER-ISOLATION-STAGE-2  all  --  anywhere             anywhere            
> RETURN     all  --  anywhere             anywhere            
> 
> Chain DOCKER-ISOLATION-STAGE-2 (2 references)
> target     prot opt source               destination         
> DROP       all  --  anywhere             anywhere            
> DROP       all  --  anywhere             anywhere            
> RETURN     all  --  anywhere             anywhere            
> 
> Chain DOCKER-USER (1 references)
> target     prot opt source               destination         
> RETURN     all  --  anywhere             anywhere            
> 
> Chain sshguard (1 references)
> target     prot opt source               destination 

I am on google cloud (GCP) Compute Engine VM if that makes a difference.

TL;DR - how can I enable opening bitcoind and lightning to the outside world using their native ports (8333, 9735) on btcpayserver-docker?
Basically, I want to see it with 8333 here:
https://bitnodes.earn.com/

Thanks for ANY help!

**Comment by NicolasDorier** [comment]
@maltokyo sorry but for now it is not possible. The reason is that all incoming connection to the full node would be considered whitelisted, which is bad for privacy.

If you want to connect to your own node, you can via the Tor link in `Server Settings / services / full node`. But this is only for private usage as it would open use to DDoS attacks.

**Comment by maltokyo** [comment]
@NicolasDorier, thank you very much! Appreciate your time.  What about for Lightning node (9735).  Shouldn't it be reachable from the internet?  How can I connect to my own node (direct link to wallet etc) with a 3rd party lightning wallet?  Or is the built-in Spark wallet the only way to do this?  (unfortunately, I am on iOS for phone, and can't find any wallet to connect with Lightning).

**Comment by NicolasDorier** [comment]
@maltokyo your lightning node should be exposed already.

**Comment by NicolasDorier** [comment]
Check https://github.com/btcpayserver/btcpayserver-doc/blob/master/LightningNetwork.md

**Comment by maltokyo** [comment]
@NicolasDorier, thank you again.  Lightning node seems not exposed.  
My server is pay.malpage.com:9735 - maybe I am looking at it or testing it the wrong way, but the port/server doesn't look exposed for me.  I pasted in the original post the ss command and the iptables output.  Does it look correct for you? Also, port 9735 is open on GCP firewall settings, in the same way as 80 and 443 for the Nginx reverse proxy, which works fine.

Any suggestions welcome!  I just can't figure it out.

**Comment by NicolasDorier** [comment]
Do `docker ps` and check lnd or clightning is running, you can see port mapping as well.

**Comment by NicolasDorier** [comment]
If you used the docker install, the ports should be exposed.

**Comment by maltokyo** [comment]
Hi @NicolasDorier 
Thanks again.  I found the error (of fix, needed for GCP).

In docker-compose.generated.yml currently for c-lightning it has

```
    ports:
    - "9735:9735"
```

This needs to be:

```
    ports:
    - "0.0.0.0:9735:9735"
```

You may want to fix this in your setup.  

Background how I found it - after reading here:
https://stackoverflow.com/questions/24281946/how-to-open-incoming-port-50070-in-firewall-google-compute-engine

I picked up the hint from here how to solve it:
https://github.com/ElementsProject/lightning/blob/master/README.md

It now works!  I think you safely change this for all the ports you expose for all services.  Anyone on the Google Cloud Platform will face this issue, with their funky network settings.

(I don't know why it just works for nginx reverse proxy though...)

**Comment by NicolasDorier** [comment]
wtf? I never had any problem in any other environment.

**Comment by NicolasDorier** [comment]
@maltokyo do you have any idea why it happened on your environment and not any other? I am worried that if I change I might break for other persons.

**Comment by NicolasDorier** [comment]
Can you tell me the content of the file `/etc/docker/daemon.json`

**Comment by maltokyo** [comment]
The thing is, on Google cloud the network setup is
Public IP -> Virtual machine IP (10.0.0.x) -> BTCPay Server Docker IPs (172.x.x.x)
So when the docker IP is listening with just `ACCEPT tcp -- anywhere 172.28.0.8 tcp dpt:9735` it only listens to incoming connections from 10.0.0.x (and not from the public IP.

When it is set to listen on 0.0.0.0 it means literally accept connections from ANYWHERE they come from.  At least on google cloud.  I think....  

```
cat /etc/docker/daemon.json
{
"log-driver": "json-file",
"log-opts": {"max-size": "5m", "max-file": "3"}
}
```

**Comment by NicolasDorier** [comment]
@maltokyo

> So when the docker IP is listening with just ACCEPT tcp -- anywhere 172.28.0.8 tcp dpt:9735 it only listens to incoming connections from 10.0.0.x (and not from the public IP.

But why does docker choose 172.28.0.8  over the public interface?

**Comment by NicolasDorier** [comment]
For example, this post show that for this user, docker listen to 0.0.0.0 by default

https://github.com/docker/for-linux/issues/370#issue-345153029

**Comment by NicolasDorier** [comment]
This article also mention it is 0.0.0.0 by default https://forums.docker.com/t/can-i-change-the-default-ip-from-0-0-0-0-when-binding/30358

**Comment by maltokyo** [comment]
All I know is that mine was not open on the public IP until I specifically added 0.0.0.0
As said, I think it is because of the multiple layers of the network on GCP..  Which is unavoidable, as you cannot choose to have less.

**Comment by maltokyo** [comment]
what is strange is, that I didnt need to change anything for Nginx.. it "just works"

**Comment by NicolasDorier** [comment]
@maltokyo  can you try to set the ip to 0.0.0.0 in the daemon.json as the links above suggests ? Maybe you won't need to specify 0.0.0.0

**Comment by maltokyo** [comment]
OK, @NicolasDorier  I tried as you suggested, but I am not sure if it worked.
Would you mind to try to connect with me?  If you can connect, the fix has worked with json file.

Here is my node address:

`02aa167aaee0cdfbfed3394e20cbaf2dca3b3b4613c3f93f23b3e4a6dddb64df91@35.211.81.213:9735`

**Comment by NicolasDorier** [comment]
it seems to work!

**Comment by maltokyo** [comment]
great!  Maybe you could add this to documentation for GCP people.  Mine currently looks like this

```
# cat /etc/docker/daemon.json
{
"log-driver": "json-file",
"log-opts": {"max-size": "5m", "max-file": "3"},
"ip" : "0.0.0.0"
}
```

**Comment by NicolasDorier** [comment]
@maltokyo it is so strange that your default is different from anybody else...

**Comment by maltokyo** [comment]
actually, this change to cat /etc/docker/daemon.json is NOT working, I can no longer receive inbound payments where a channel is a setup.
I changed again docker-compose.generated.yml and it works again.  So I still suggest changing that.

**Comment by NicolasDorier** [comment]
That's strange, I tried to connect to your node from my node and it worked.

**Comment by maltokyo** [comment]
Well a friend could open a channel with me while I had the changes in yaml file but not without it. Maybe it was another issue? If you like I can change it back, and you could attempt to open a channel with me? I’m not sure of any other way to verify this.

**Comment by NicolasDorier** [comment]
I did not try to open a channel, just connecting to your node.

**Comment by maltokyo** [comment]
I moved away from GCP, as this was never resolved.  So closing
